ckanext/switzerland/helpers.py

```diff
@@ -470,3 +487,105 @@ def get_cookie_law_url():
 
 def get_cookie_law_id():
     return tk.config.get("ckanext.switzerland.cookie_law_id", "")
+
+
+def _get_datetime_from_isoformat_string(dt_string):
+    """Get a datetime object from an isoformat string, and handle empty values
+    and malformatted strings.
+
+    Returns the datetime object if possible, False if it couldn't get one.
+    """
+    if dt_string in [None, ""]:
+        return False
+    try:
+        return datetime.fromisoformat(dt_string)
+    except ValueError:
+        log.warning(f"Error getting a datetime from isoformat string: {dt_string}")
+        return False
+
+
+def convert_datetimes_for_display(dataset_or_resource_dict):
+    """Converts all datetimes in a dataset or resource to UTC, without time zone info
+    attached. CKAN's "automatic-local-datetime" HTML class and JS helper will then
+    display them in the user's local time zone.
+
+    CKAN stores all datetimes as UTC by default. For our custom datetime fields, we use
+    the server time zone (Europe/Zurich), so they have to be converted to UTC.
+    """
+    for field in CUSTOM_DATETIME_FIELDS:
+        dt = _get_datetime_from_isoformat_string(dataset_or_resource_dict.get(field))
+        if dt is False:
+            continue
+
+        dt_zh = dt.replace(tzinfo=ZURICH)
+        dt_utc = dt_zh.astimezone(UTC)
+        dataset_or_resource_dict[field] = dt_utc.replace(tzinfo=None).isoformat()
+
+    if dataset_or_resource_dict.get("extras"):
+        for extra in dataset_or_resource_dict["extras"]:
+            if extra["key"] in CUSTOM_DATETIME_FIELDS:
+                dt = _get_datetime_from_isoformat_string(extra["value"])
+                if dt is False:
+                    continue
+
+                dt_zh = dt.replace(tzinfo=ZURICH)
+                dt_utc = dt_zh.astimezone(UTC)
+                extra["value"] = dt_utc.replace(tzinfo=None).isoformat()
+
+    if dataset_or_resource_dict.get("resources"):
+        for resource in dataset_or_resource_dict["resources"]:
+            convert_datetimes_for_display(resource)
+
+
+def convert_datetimes_for_api(dataset_or_resource_dict):
+    """Calculates the time of a datetime in the Europe/Zurich time zone and outputs the
+    value as isoformat, with time zone info.
+
+    CKAN stores all datetimes as UTC by default, so they have to be converted to
+    Europe/Zurich and have the time zone info added. For our custom datetime fields, we
+    use the server time zone, so they are already in Europe/Zurich. We just have to add
+    the time zone info.
+    """
+    for field in CKAN_DATETIME_FIELDS:
+        dt = _get_datetime_from_isoformat_string(dataset_or_resource_dict.get(field))
+        if dt is False:
+            continue
+
+        dt_utc = dt.replace(tzinfo=UTC)
+        dt_zh = dt_utc.astimezone(ZURICH)
```

Comment:
Same as above. I don't think it should hurt us a lot, but as it is more than one parameter we are changing every time, this might be worth checking?

Reply:
Same answer as above. :)